File: frameworks/go/gin/Implementation/contract.go
package implementation

import (
	"net/http"
	"runtime"
	"runtime/debug"
	"strings"

	"github.com/gin-gonic/gin"
	ginjson "github.com/gin-gonic/gin/codec/json"
)

// contractRoutes answers /health and /__meta, which the contract asks of every framework
// outside the corpus.
func contractRoutes(r *gin.Engine) {
	// The payloads are loaded before the server listens, so a server that answers has them.
	r.GET("/health", func(c *gin.Context) { c.String(http.StatusOK, "ok") })

	version := ginVersion()
	r.GET("/__meta", func(c *gin.Context) {
		c.JSON(http.StatusOK, gin.H{
			"framework": "Gin",
			"version":   version,
			"runtime":   runtime.Version(),
			"adapter":   "net/http",
			// The codec gin's build tags selected, as gin records it.
			"serializer": ginjson.Package,
			// Go sets this from the container's CPU quota, so it says how many threads run Go code.
			"gomaxprocs": runtime.GOMAXPROCS(0),
		})
	})
}

// ginVersion is the gin module the build resolved, without the v Go puts in front of it.
func ginVersion() string {
	if info, ok := debug.ReadBuildInfo(); ok {
		for _, dep := range info.Deps {
			if dep.Path == "github.com/gin-gonic/gin" {
				return strings.TrimPrefix(dep.Version, "v")
			}
		}
	}
	return strings.TrimPrefix(gin.Version, "v")
}

File: frameworks/rust/poem/Implementation/routes/contract.rs
use poem::endpoint::make_sync;
use poem::web::Json;
use poem::{Route, get};
use serde::Serialize;

use crate::Payloads;

/// /__meta: what ran, recorded on the result row and never checked. build.rs reads the versions
/// from Cargo.lock and rustc when the binary is built.
#[derive(Serialize)]
struct Meta {
    framework: &'static str,
    version: &'static str,
    runtime: &'static str,
    serializer: &'static str,
}

const META: Meta = Meta {
    framework: "poem",
    version: env!("RB_POEM_VERSION"),
    runtime: env!("RB_RUSTC_VERSION"),
    serializer: concat!("serde_json ", env!("RB_SERDE_JSON_VERSION")),
};

/// /health and /__meta, which the contract asks of every framework outside the corpus.
pub fn add(route: Route, _: &'static Payloads) -> Route {
    route
        // The payloads are loaded before the server starts, so a server that answers has them.
        .at("/health", get(make_sync(|_| "ok")))
        .at("/__meta", get(make_sync(|_| Json(META))))
}
